src/main/java/net/neoforged/neoforge/capabilities/color/IColorable.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities.color;
+
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.item.DyeColor;
+import net.minecraft.world.item.ItemStack;
+import net.neoforged.neoforge.common.Tags;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents a colorable object in-game.
+ */
+public interface IColorable {
+    /**
+     * Suggests to an applicator whether applying color to the target should consume some quantity
+     * of dye/ink/etc. Implementation of consumption is left as an exercise to the applicator.
+     *
+     * @return True if dye should be considered consumed on success; false otherwise.
+     */
+    default boolean consumesDye(@Nullable LivingEntity entity, ItemStack source) {
+        return source.is(Tags.Items.DYES);
```

Comment:
This makes it a bit hard for items that are not dyes to determine if the action should have a cost. For example, an FE-powered dye applicator that's loaded with colour charges wouldn't want to be a dye item, but it would like to use charges when used on a sheep but not when used on a combination lock. Better leave source-specific decisions to the source item and keep this one all about the target.

Reply:
This is for situations where you're actually using VANILLA dye and need to suggest to the vanilla logic that it should use up a dye item. Blame signs.

Reply:
Then why not put the item check into those two calling locations that are only intended for vanilla items?

`if (colorable.consumesDye(player, stack.copy()) && stack.is(Tags.Items.DYES)) {`

That way any non-vanilla item can consume if they want to. At the moment, `consumesDye()` is pretty useless and could be called `isVanillaDye()` instead. A non-vanilla item that can dye stuff would not have any reason to call it, as it already knows that all vanilla IColorables (and modded ones using the default) will always return false.

I mean, one could do something like:

```java
boolean consumes_me = colorable.consumesDye(player, stack.copy());
if (!consumes_me) {
  boolean consumes_dye = colorable.consumesDye(player, STACK_OF_BONEMEAL.copy());
  if (consumes_dye) {
    boolean consumes_non_dye = colorable.consumesDye(player, STACK_OF_STONE.copy());
    if (!consumes_non_dye) {
      consumes_me = true; // consumes only dyes
    } else {
      consumes_me = false; // seems to specially handle me
    }
  } else {
    consumes_me = false; // doesn't consume dye and doesn't want to consume me
  }
}
```

Which will catch IColorables that use the default impl, ones that always return true, ones that always return false, and, to some extent, those that know the calling item and handle it in a special way. And is very ugly.

Making the calls in the form `consumesDye() && isConsumablDye()` is much more elegant.

---

Assuming modded items inherit `IDyeItemExtension` and don't want to copy the whole `vanillaUseOn()` (which in many cases is perfectly fine as is, even though it's intended for vanilla items), adding this to the interface would make sense in my opinion:

```java
default boolean isConsumableDye(ItemStack stack) {
    return stack.is(Tags.Items.DYES);
}
```

I personally would also put the `stack.consume(1, player);` into a method, allowing modded items to use the good default impls for the interaction even when using another way to store charges. But that is personal preference. I hate having to copy&paste whole methods from superclasses just to change one line. That stuff tends to break silently when porting to a new version as the copy very likely desyncs but still compiles well. I had to fix way too many bugs of that kind way too often...